Brille integration

rebeccafair opened the pull request:
Note should be merged after https://github.com/pace-neutrons/Euphonic/pull/104

To do once https://github.com/pace-neutrons/Euphonic/pull/104 has been merged and released:
- [x] Update changelog with correct release Euphonic version
- [x] Update min_requirements.txt with correct release Euphonic version
- [x] Revert .github/workflows/tests.yml to test against release Euphonic and not 95_brille_integration branch

rebeccafair committed:
Ensure qpts are contiguous as required by Brille

The T operation only changes strides, not the data layout

rebeccafair committed:
Add BrilleInterpolator to type hints

rebeccafair committed:
Add initial Brille test

rebeccafair committed:
Clone Euphonic Brille branch

rebeccafair committed:
Fix extras specification in apply_requirements

Previously no extras ended up with the command
'python -mpip install .[]', '[]' is invalid, so make sure
in this case it's set to an empty string instead

rebeccafair committed:
Fix cloning in apply_requirements

rebeccafair committed:
Add test to check kwargs passed

rebeccafair committed:
Update changelog

rebeccafair committed:
Update changelog and min_requirements with Euphonic version

mducle commented on .github/workflows/tests.yml:
```diff
@@ -34,12 +34,17 @@ jobs:
           python apply_requirements.py --extras phonopy_reader ${{ matrix.euphonic_version }}
           python run_tests.py -m "phonopy_reader" --coverage "coverage_phonopy_reader.xml"
+      - name: Install Euphonic with Brille extra and run tests
+        shell: bash -l {0}
```
I think you can probably put a section

```yaml
defaults:
  run:
    shell: bash -l {0}
```
At the top of the file since all the run steps use `bash` in this file.

rebeccafair replied:
Done!

mducle commented on euphonic_sqw_models/euphonic_wrapper.py:
```diff
@@ -154,10 +145,13 @@ def __init__(
+        for int_kwarg in int_kwargs:
+            if int_kwarg in calc_modes_kwargs:
+                calc_modes_kwargs[int_kwarg] = \
+                    int(calc_modes_kwargs[int_kwarg])
```
It might not be as pretty, but I think you can do this in one line:

```python
calc_modes_kwargs.update({k:int(v) for k, v in ['threads', 'n_threads'] if k in calc_modes_kwargs})
```

rebeccafair replied:
I think readability might be better here, so I've left the explicit if & for in, but I did reduce `int_kwarg` to `ikw` so the assignment now fits on one line.

mducle commented on test/euphonic_horace_test.py:
```diff
@@ -386,3 +384,59 @@ def test_optimum_chunk_mocked_psutil_virtualmemory(self, mocker):
+        fname = get_expected_output_filename('quartz', opt_dict)
+        # Ignore vals below 25 meV - hard to test
+        expected_w, expected_sf, idx_ignore = get_expected_w_sf(
+            fname, lim=25)
```
This is more a comment about the accuracy of the Brille interpolation rather than the test or implementation, but it's a bit concerning to me that you need to set a limit of 25 meV, and moreover even then the absolute tolerance needed to pass in line 412 below is 3 meV. These numbers just seem quite large to me, implying that the Brille interpolation is not really that good. I guess it stems from using quite a coarse grid of `grid_npts=1000`?

I think the maximum frequency in quartz is around 150 meV, so the "low" frequency limit here is about 15% of the max, and the tolerance is about 2% of max... I guess it's maybe not unreasonable but I had thought Brille would be better... What if you used say, `grid_npts=5000`?

Also, what is the length of `idx_ignore` compared to `expected_w`? If it's small (say a few percent) then I guess it's ok.

rebeccafair replied:
Part of it is using a coarse grid yes, but I think because quartz has a lot of degenerate frequencies perhaps brille has a hard time tracking the modes sometimes, and we end up with the odd frequency or structure factor which is a bit off, this works out ok for large numbers of q-points but makes it hard to test numerically. I've run `euphonic-brille-convergence quartz.castep_bin --brille-npts 1000 5000 10000` and here are the results:
![image](https://user-images.githubusercontent.com/18187336/217539793-5230f448-c017-4ca0-a6c5-3e2a490f61d4.png)
So overall it's pretty good, but there is the odd issue with frequencies/structure factors at certain points.

I used quartz because I wanted to have a brille-extra-only test (rather than using NaCl and having to rely on the Phonopy dependencies as well). Here is a similar run for NaCl:
![image](https://user-images.githubusercontent.com/18187336/217540455-337ed410-b1a0-4348-945c-ab2ea27ab559.png)
So the frequencies are much better but there's still a bit of discrepancy in the intensities. I could use NaCl instead but it's still just a case of fiddling with the numbers until the tests pass...

When `lim=25`, `idx_ignore` has 41 values, and `expected_w` has `8*27 = 216` values, so about 20%...

mducle replied:
I think it's ok to leave quartz as the test case. In fact I think the test itself is fine, it was just the thresholds which surprised me a bit. Given that it seems to be skewed by a few outliers, would it make sense to use the mean absolute frequency residual as the criteria rather than `allclose`?

rebeccafair replied:
Good idea, done in [1e05b5d](https://github.com/pace-neutrons/euphonic_sqw_models/pull/11/commits/1e05b5d50359a92cd6f232e21327dabe73b93e4b).

I have also used the structure factor residuals. It is enough to fail if the proper parameters don't get passed (e.g if I don't use the correct temperature) so I think that's good enough to tell us if something goes wrong.

mducle reviewed (approved):
Looks good. Just a couple of minor comments, but I don't think anything needs changing...

rebeccafair committed:
Updates from review:

- Use defaults in Github actions
- Reduce number of lines in threads/n_threads int conversion

rebeccafair committed:
Use mean residual in tests due to outliers

rebeccafair committed:
Test against release Euphonic
